File: hw2/grades/test/grades/calculator_test.exs
defmodule Grades.CalculatorTest do
  use ExUnit.Case
  alias Grades.Calculator

  describe "percentage_grade/1" do
    test "percent_grade_tests" do
      assert 85 ==
               Calculator.percentage_grade(%{
                 homework: [0.8],
                 labs: [1, 1, 1],
                 midterm: 0.70,
                 final: 0.9
               })
	  assert 18 ==
               Calculator.percentage_grade(%{
                 homework: [],
                 labs: [],
                 midterm: 0.3,
                 final: 0.4
               })
    end
	
	test "letter_grade_tests" do
      assert "EIN" ==
               Calculator.letter_grade(%{
                 homework: [],
                 labs: [],
                 midterm: 0,
                 final: 0
               })
	  assert "A+" ==
               Calculator.letter_grade(%{
                 homework: [0.9],
                 labs: [1, 1, 1],
                 midterm: 0.8,
                 final: 0.9
               })
	  assert "A" ==
               Calculator.letter_grade(%{
                 homework: [0.8],
                 labs: [1, 1, 1],
                 midterm: 0.7,
                 final: 0.9
               })
	  assert "A-" ==
               Calculator.letter_grade(%{
                 homework: [0.8],
                 labs: [1, 0.9, 0.7],
                 midterm: 0.7,
                 final: 0.9
               })
      assert "B+" ==
               Calculator.letter_grade(%{
               homework: [0.7],
               labs: [0.9, 0.9, 0.9],
               midterm: 0.7,
               final: 0.8
             })
      assert "B" ==
             Calculator.letter_grade(%{
               homework: [0.8],
               labs: [0.7, 0.6, 0.7],
               midterm: 0.6,
               final: 0.7
             })
      assert "C+" ==
             Calculator.letter_grade(%{
               homework: [0.7],
               labs: [0.7, 0.6, 0.7],
               midterm: 0.6,
               final: 0.7
             })
      assert "C" ==
             Calculator.letter_grade(%{
               homework: [0.7],
               labs: [0.7, 0.5, 0.7],
               midterm: 0.6,
               final: 0.5
             })
      assert "D+" ==
             Calculator.letter_grade(%{
               homework: [0.5],
               labs: [0.7, 0.5, 0.7],
               midterm: 0.6,
               final: 0.5
             })
      assert "D" ==
             Calculator.letter_grade(%{
               homework: [0.5],
               labs: [0.7, 0.2, 0.7, 0.5],
               midterm: 0.6,
               final: 0.5
             })
      assert "E" ==
             Calculator.letter_grade(%{
               homework: [0.5],
               labs: [0.4, 0.4, 0.4],
               midterm: 0.5,
               final: 0.5
             })
      assert "F" ==
             Calculator.letter_grade(%{
               homework: [0.4],
               labs: [0.4, 0.4, 0.4],
               midterm: 0.5,
               final: 0.3
             })	   
    end
	
	test "numeric_grade_tests" do
      assert 0 ==
               Calculator.numeric_grade(%{
                 homework: [],
                 labs: [],
                 midterm: 0,
                 final: 0
               })
	  assert 10 ==
               Calculator.numeric_grade(%{
                 homework: [0.9],
                 labs: [1, 1, 1],
                 midterm: 0.8,
                 final: 0.9
               })
	  assert 9 ==
               Calculator.numeric_grade(%{
                 homework: [0.8],
                 labs: [1, 1, 1],
                 midterm: 0.7,
                 final: 0.9
               })
	  assert 8 ==
               Calculator.numeric_grade(%{
                 homework: [0.8],
                 labs: [1, 0.9, 0.7],
                 midterm: 0.7,
                 final: 0.9
               })
      assert 7 ==
               Calculator.numeric_grade(%{
               homework: [0.7],
               labs: [0.9, 0.9, 0.9],
               midterm: 0.7,
               final: 0.8
             })
      assert 6 ==
             Calculator.numeric_grade(%{
               homework: [0.8],
               labs: [0.7, 0.6, 0.7],
               midterm: 0.6,
               final: 0.7
             })
      assert 5 ==
             Calculator.numeric_grade(%{
               homework: [0.7],
               labs: [0.7, 0.6, 0.7],
               midterm: 0.6,
               final: 0.7
             })
      assert 4 ==
             Calculator.numeric_grade(%{
               homework: [0.7],
               labs: [0.7, 0.5, 0.7],
               midterm: 0.6,
               final: 0.5
             })
      assert 3 ==
             Calculator.numeric_grade(%{
               homework: [0.5],
               labs: [0.7, 0.5, 0.7],
               midterm: 0.6,
               final: 0.5
             })
      assert 2 ==
             Calculator.numeric_grade(%{
               homework: [0.5],
               labs: [0.7, 0.2, 0.7, 0.5],
               midterm: 0.6,
               final: 0.5
             })
      assert 1 ==
             Calculator.numeric_grade(%{
               homework: [0.5],
               labs: [0.4, 0.4, 0.4],
               midterm: 0.5,
               final: 0.5
             })
      assert 0 ==
             Calculator.numeric_grade(%{
               homework: [0.4],
               labs: [0.4, 0.4, 0.4],
               midterm: 0.5,
               final: 0.3
             })	   
    end
  end
end

File: hw2-midnight/grades_java (optional)/test/CalculatorTest.java
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

class CalculatorTest {

	@Test
	void percentGradeTests() {
		assertEquals(85, Calculator.percentageGrade(new float[] {0.8f}, new float[] {1f,  1f,  1f}, 0.7f, 0.9f));
		assertEquals(18, Calculator.percentageGrade(new float[0], new float[0], 0.3f, 0.4f));
	}
	
	@Test
	void letterGradeTests() {
		assertEquals("EIN", Calculator.letterGrade(new float[0], new float[0], 0f, 0f)); 
		assertEquals("A+", Calculator.letterGrade(new float[] {0.9f}, new float[] {1f,  1f,  1f}, 0.8f, 0.9f));
		assertEquals("A", Calculator.letterGrade(new float[] {0.8f}, new float[] {1f,  1f,  1f}, 0.7f, 0.9f));
		assertEquals("A-", Calculator.letterGrade(new float[] {0.8f}, new float[] {1f,  0.9f,  0.7f}, 0.7f, 0.9f));
		assertEquals("B+", Calculator.letterGrade(new float[] {0.7f}, new float[] {0.9f,  0.9f,  0.7f}, 0.7f, 0.8f));
		assertEquals("B", Calculator.letterGrade(new float[] {0.8f}, new float[] {0.7f,  0.6f,  0.7f}, 0.6f, 0.7f));
		assertEquals("C+", Calculator.letterGrade(new float[] {0.7f}, new float[] {0.7f,  0.6f,  0.7f}, 0.6f, 0.7f));
		assertEquals("C", Calculator.letterGrade(new float[] {0.7f}, new float[] {0.7f,  0.5f,  0.7f}, 0.6f, 0.5f));
		assertEquals("D+", Calculator.letterGrade(new float[] {0.5f}, new float[] {0.7f,  0.5f,  0.7f}, 0.6f, 0.5f));
		assertEquals("D", Calculator.letterGrade(new float[] {0.5f}, new float[] {0.7f,  0.2f,  0.7f, 0.5f}, 0.6f, 0.5f)); //!!
		assertEquals("E", Calculator.letterGrade(new float[] {0.5f}, new float[] {0.4f,  0.4f,  0.4f}, 0.5f, 0.5f));
		assertEquals("F", Calculator.letterGrade(new float[] {0.4f}, new float[] {0.4f,  0.4f,  0.4f}, 0.5f, 0.3f));
	}
	
	@Test
	void numericGradeTests() {
		assertEquals(0, Calculator.numericGrade(new float[0], new float[0], 0f, 0f)); 
		assertEquals(10, Calculator.numericGrade(new float[] {0.9f}, new float[] {1f,  1f,  1f}, 0.8f, 0.9f));
		assertEquals(9, Calculator.numericGrade(new float[] {0.8f}, new float[] {1f,  1f,  1f}, 0.7f, 0.9f));
		assertEquals(8, Calculator.numericGrade(new float[] {0.8f}, new float[] {1f,  0.9f,  0.7f}, 0.7f, 0.9f));
		assertEquals(7, Calculator.numericGrade(new float[] {0.7f}, new float[] {0.9f,  0.9f,  0.7f}, 0.7f, 0.8f));
		assertEquals(6, Calculator.numericGrade(new float[] {0.8f}, new float[] {0.7f,  0.6f,  0.7f}, 0.6f, 0.7f));
		assertEquals(5, Calculator.numericGrade(new float[] {0.7f}, new float[] {0.7f,  0.6f,  0.7f}, 0.6f, 0.7f));
		assertEquals(4, Calculator.numericGrade(new float[] {0.7f}, new float[] {0.7f,  0.5f,  0.7f}, 0.6f, 0.5f));
		assertEquals(3, Calculator.numericGrade(new float[] {0.5f}, new float[] {0.7f,  0.5f,  0.7f}, 0.6f, 0.5f));
		assertEquals(2, Calculator.numericGrade(new float[] {0.5f}, new float[] {0.7f,  0.2f,  0.7f, 0.5f}, 0.6f, 0.5f)); //!!
		assertEquals(1, Calculator.numericGrade(new float[] {0.5f}, new float[] {0.4f,  0.4f,  0.4f}, 0.5f, 0.5f));
		assertEquals(0, Calculator.numericGrade(new float[] {0.4f}, new float[] {0.4f,  0.4f,  0.4f}, 0.5f, 0.3f));
	}
}
